### DISCUSSION TITLE
`time '...'` literal cannot convert values greater than 24 hours

### DISCUSSION THREAD

**Discussion by bgrainger** [planetscale/discussion#369] [Bug Reports] [answered]
MySQL Community Edition (8.0.31) supports the following range for `TIME` columns:

> `TIME` values may range from `'-838:59:59'` to `'838:59:59'`.

https://dev.mysql.com/doc/refman/8.0/en/time.html

PlanetScale (8.0.23) doesn't permit as large a range.

```sql
create table times(tm time);
insert into times values(time '23:59:59.999');
insert into times values(time '-23:59:59.999');

-- Error Code: 1105. incorrect TIME value: '24:00:00'
insert into times values(time '24:00:00');
-- Error Code: 1105. incorrect TIME value: '-24:00:00'
insert into times values(time '-24:00:00');
```

Expectation is that PlanetScale would accept a range of values at least as large as MySQL, and certainly greater than 24 hours.

I did not find any mention of this compatibility difference at https://planetscale.com/docs/reference/mysql-compatibility.

**Comment by orware** [comment, marked as answer]
Hello @bgrainger,

I have run some tests on my end, and I don't think this is a PlanetScale issue since using values within the larger range that the `TIME` type allows should be possible. 

I will still continue the investigation a bit further on our end though since within a local MySQL 8.0.28 instance I didn't encounter the same issue when using your examples as-is so I'm not 100% sure why using `time '-838:59:58'` behaves differently than `time('-838:59:58')` but the first seems to be limited to interpreting values within the 24-hour range although a reference explaining this difference in the MySQL documentation didn't turn up right away when I ran a search.

Based on what's contained in this [Stack Overflow post](https://stackoverflow.com/questions/39259910/why-is-mysqls-maximum-time-limit-8385959) it doesn't appear that the range of `TIME` was changed within a recent MySQL 8.0.x release.

In my own tests within our web console using the examples you provided, I did receive the same error if I used `time` as you did in your example, but if I removed that, it worked as you can see below:
![image](https://user-images.githubusercontent.com/931650/211909133-c3a4e271-9191-4339-93e8-48bdfff126ce.png)

My explanation for the behavior here, is that when you include the `time` part, something may be interpreting things incorrectly rather than how the actual `TIME()` [function](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_time) would process things.

Here's an additional example showing that test:
![image](https://user-images.githubusercontent.com/931650/211912233-b38ac305-13eb-4f90-80ad-c4951a11a5bc.png)

If I adjust your example to be wrapped in the `TIME()` function it also works correctly:
![image](https://user-images.githubusercontent.com/931650/211912439-021cb1e5-94f3-47e9-84ba-efc86ef59bfb.png)

So switching to this alternative approach should allow those larger values to be inserted correctly for your needs.

**Reply by orware** [reply]
@bgrainger,

To follow-up with the tests and information above, after I shared the issue internally, it sounds like our Vitess layer does have some room for improvement when interpreting the TIME literal option:
https://github.com/vitessio/vitess/blob/main/go/vt/sqlparser/parse_date.go#L44-L47

```
// ParseTime is right now only excepting on specific
// time format and doesn't accept all formats MySQL accepts.
// Can be improved in the future as needed.
```

And based on your tests it seems that Vitess is limiting the TIME literal approach to the `-23:59:59` to `23:59:59` range rather than the larger one that the `TIME` type actually supports.

But if you use the `TIME()` function, the query is able to pass through the Vitess parser for that and allows the underlying MySQL 8 instance to process things as it normally would, which is why using that option doesn't run into the same issue.

**Comment by bgrainger** [comment]
Thanks for the reply. 

I was deliberately using the `TIME '..'` literal (https://dev.mysql.com/doc/refman/8.0/en/date-and-time-literals.html) not the `TIME()` function (https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_time) but see that I may have incorrectly conflated errors from using the literal syntax with problems in using the `TIME` data type itself. (If so, the summary I used as the discussion topic is wrong.)

Let me perform some more detailed experiments and update you here.
